ValveResourceFormat/IO/MaterialExtract.cs

```diff
@@ -0,0 +1,482 @@
+        },
+
+        ["sky"] = new()
+        {
+            ["g_tSkyTexture"] = new[] { (Channel.All, "SkyTexture") },
+        }
+    };
+
+    public static readonly Dictionary<string, string> CommonTextureSuffixes = new()
+    {
+        { "TextureDetailMask", "_detailmask" },
+        { "TextureDiffuseWarpMask", "_diffusemask" },
+        { "TextureMetalnessMask", "_metalnessmask" },
+        { "TextureSelfIllumMask", "_selfillummask" },
+
+        { "TextureSpecularMask", "_specmask" },
+        { "TextureRimMask", "_rimmask" },
+        { "TextureTintByBaseMask", "_basetintmask" },
+        { "TextureSpecularExponent", "_specexp" },
+
+        { "TextureColor", "_color" },
+        { "TextureNormal", "_normal" },
+        { "TextureRoughness", "_rough" },
+        { "TextureMetalness", "_metal" },
+        { "TextureAmbientOcclusion", "_ao" },
+        { "TextureReflectance", "_refl"},
+    };
+
+    private static bool IsDefaultTexture(string textureFileName)
+        => textureFileName.StartsWith("materials/default/", StringComparison.OrdinalIgnoreCase);
+
+    private readonly Material material;
+    private readonly ResourceEditInfo editInfo;
+    private readonly IFileLoader fileLoader;
+
+    public MaterialExtract(Material material, ResourceEditInfo editInfo, IFileLoader fileLoader)
+    {
+        this.material = material;
+        this.editInfo = editInfo;
+        this.fileLoader = fileLoader;
+    }
+
+    public MaterialExtract(Resource resource, IFileLoader fileLoader = null)
+        : this((Material)resource.DataBlock, resource.EditInfo, fileLoader)
+    {
+    }
+
+    public ContentFile ToContentFile()
+    {
+        var vmat = new ContentFile
+        {
+            Data = Encoding.UTF8.GetBytes(ToValveMaterial())
+        };
+
+        foreach (var (type, filePath) in material.TextureParams)
+        {
+            var texture = fileLoader?.LoadFile(filePath + "_c");
+            if (texture is null)
+            {
+                continue;
+            }
+
+            var images = GetInputImagesForTexture(type, filePath, true, material.IntParams, material.VectorParams);
+            var vtex = new TextureExtract(texture).ToMaterialMaps(images);
+
+            vmat.ExternalRefsHandled.Add(filePath + "_c", vtex);
+            vmat.SubFilesAreExternal = true;
+            vmat.SubFiles.AddRange(vtex.SubFiles);
+        }
+
+        return vmat;
+    }
+
+    public static string OutTextureName(string texturePath, bool keepOriginalExtension, string desiredSuffix = null)
+    {
+        if (!texturePath.EndsWith(".vtex", StringComparison.OrdinalIgnoreCase))
+        {
+            throw new ArgumentException($"{texturePath} must have .vtex", nameof(texturePath));
+        }
+        texturePath = Path.ChangeExtension(texturePath, null); // strip .vtex
+        texturePath = Path.ChangeExtension(texturePath, null); // strip .generated (if present)
```

Comment:
`if (texturePath.EndsWith(".generated", StringComparison.OrdinalIgnoreCase))`

Reply:
Even if `Path.ChangeExtension` does the same check and will leave it unchanged?

Reply:
ChangeExtension will strip anything if the file name has a dot in it.